### DISCUSSION TITLE
rabbitmq_aws does not escape forward slashes in path, service and region

### DISCUSSION THREAD

**Discussion by jovilark** [rabbitmq/rabbitmq-server#5242] [Other]
Hi RabbitMQ,

When creating a request, including a forward slash character in the path, service or region results in the rest of the URI being snipped early.

examples:

set_region("us-west-2")
rabbitmq_aws:get("lambda", "/2015-03-31/functions/FunctionName"),
URI constructed in request:
expected: lambda.us-west-2.amazonaws.com/2015-03-31/functions/FunctionName
actual: lambda.us-west-2.amazonaws.com/2015-03-31

set_region("us-west-2")
rabbitmq_aws:get("lam/bda", "/2015-03-31/functions/FunctionName"),
URI constructed in request:
expected: lam/bda.us-west-2.amazonaws.com/2015-03-31/functions/FunctionName
actual: lam

set_region("us-west/-2"),
rabbitmq_aws:get("lambda", "/2015-03-31/functions/FunctionName"),
URI constructed in request:
expected: lam/bda.us-west-2.amazonaws.com/2015-03-31/functions/FunctionName
actual: lambda.us-west

I haven't been able to identify exactly where this truncation occurs with my unfamiliarity with Erlang, I apologise.
Thank you!

**Comment by michaelklishin** [comment]
I'm not convinced that automatic encoding of slashes in path segments always should be done by libraries.

How does the affect RabbitMQ users?

**Comment by jovilark** [comment]
Using the RabbitMQ AWS library, I would like to make a request to a path that has forward slashes. How can this be done otherwise?

**Comment by michaelklishin** [comment]
@jovilark encode them as path segments in your own code. IIRC path segments of URIs are percent-encoded (but different URI components are encoded differently).

**Comment by jovilark** [comment]
@michaelklishin Do you mind showing an example ?
I've tried a few ways to do this with no success.
I percent encoded the entire path
/My/Path -> %2FMy%2FPath
/My/Path -> /My%2FPath

Neither work

**Reply by jovilark** [reply]
I get 'bad gateway' with either.

**Reply by michaelklishin** [reply]
Only the values that are path segments must be %-encoded.

If an ID is `a/b` and the path is `/ids/{id}` then the result would be `/ids/a%2Fb`.

**Reply by jovilark** [reply]
This is not supported by the current code. Any forward slash that is not at the beginning of the path snips the path entirely, like in my example.
In your example the path would be myhost.com/ids

**Reply by jovilark** [reply]
This is the problem I am talking about

**Reply by jovilark** [reply]
rabbitmq_aws:get("lambda", "/2015-03-31/functions%2FFunctionName")
library tries to connect to URI: lambda.us-west-2.amazonaws.com/2015-03-31

**Reply by michaelklishin** [reply]
@jovilark like I said earlier, libraries like this do not have to do the escaping for you. You can do it and provide an escaped path.

This is a very simplistic client, by design.

**Comment by michaelklishin** [comment]
In general, if a library accepts a URI path, you can be sure that you are responsible for the escaping part. The library cannot possibly know what should be escaped and what should not.

Libraries that would like to provide automatic escaping have no choice but to accept path segments as a list, which is less convenient in its own way.

**Comment by jovilark** [comment]
I still am having trouble understanding the solution to the issue I am having, or implementing what you are saying.
Could you please provide an example of how I can call the path I am trying to call?
What path do I provide to the rabbitmq_aws get function?

**Reply by michaelklishin** [reply]
The solution is to

1. Put together a list of path segments some of which need escaping, or break down your path into segments: `a/b/c/d` => `[a, b, c, d]`
2. Percent-escaping those elements of the path that need them (or maybe all of them) using  [`uri_string:quote/1`](https://www.erlang.org/doc/man/uri_string.html#quote-1) or another suitable `uri_string` function. As explained earlier, different URI components use different encoding scheme, so the module uses maps and not strings.
3. Joining those elements back together with a `/`
4. Providing the result to the library

This problem is no different from correct path encoding in any other programming
language.
